crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4185,6 +4301,89 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn comparison_tuples() -> anyhow::Result<()> {
+        let mut db = setup_db();
+        db.write_dedented(
+            "src/a.py",
+            r#"
+            def int_instance() -> int: ...
+
+            eq1 = (1, (2, "s")) == (1, (2, "s"))
+            eq2 = (1, 2) == (1, 2, 3)
+            eq3 = (1, "s") == (1, 2)
+            eq4 = (int_instance(),) == (1,)
+
+            not_eq1 = (1, (2, "s")) != (1, (2, "s"))
+            not_eq2 = (1, 2) != (1, 2, 3)
+
+            lt1 = (1,) < (1, 2) < (1, 3)
+            lt2 = (1, 2) < (1, 2)
+            lte = (1,) <= (1,) <= (1, 2)
+
+            gt1 = (1, 3) > (1, 2) > (1,)
+            gt2 = (1, 2) > (1, 2)
+            gte = (1, 2) >= (1,) >= (1,)
+            "#,
+        )?;
+
+        assert_public_ty(&db, "src/a.py", "eq1", "Literal[True]");
+        assert_public_ty(&db, "src/a.py", "eq2", "Literal[False]");
+        assert_public_ty(&db, "src/a.py", "eq3", "@Todo"); // TODO: It should be `Unknown`
+        assert_public_ty(&db, "src/a.py", "eq4", "@Todo"); // TODO: It should be `bool`
+
+        assert_public_ty(&db, "src/a.py", "not_eq1", "Literal[False]");
+        assert_public_ty(&db, "src/a.py", "not_eq2", "Literal[True]");
+
+        assert_public_ty(&db, "src/a.py", "lt1", "Literal[True]");
+        assert_public_ty(&db, "src/a.py", "lt2", "Literal[False]");
+        assert_public_ty(&db, "src/a.py", "lte", "Literal[True]");
+
+        assert_public_ty(&db, "src/a.py", "gt1", "Literal[True]");
+        assert_public_ty(&db, "src/a.py", "gt2", "Literal[False]");
+        assert_public_ty(&db, "src/a.py", "gte", "Literal[True]");
+
+        db.write_dedented(
+            "src/b.py",
+            r#"
+            def int_instance() -> int: ...
+
+            in1 = (1, 2) in ((1, 2), "123")
+            in2 = (1, 2) in (1, 2, 3)
+            in3 = (1,) in ((int_instance(),),)
+
+            not_in1 = (1, 2) not in ((1, 2), "123")
+            not_in2 = (1, 2) not in (1, 2, 3)
+            not_in3 = (1,) not in ((int_instance(),),)
+
+            is1 = (1, 2) is (1, 2)
+            is2 = (1, 2) is ("a", "b")
+            is3 = (1,) is (1, 2)
+
+            is_not1 = (1, 2) is not (1, 2)
+            is_not2 = (1, 2) is not ("a", "b")
+            is_not3 = (1,) is not (1, 2)
+            "#,
+        )?;
+
+        assert_public_ty(&db, "src/b.py", "in1", "Literal[True]");
+        assert_public_ty(&db, "src/b.py", "in2", "bool");
+        assert_public_ty(&db, "src/b.py", "in3", "bool");
+
+        assert_public_ty(&db, "src/b.py", "not_in1", "Literal[False]");
+        assert_public_ty(&db, "src/b.py", "not_in2", "bool");
+        assert_public_ty(&db, "src/b.py", "not_in3", "bool");
+
+        assert_public_ty(&db, "src/b.py", "is1", "bool");
+        assert_public_ty(&db, "src/b.py", "is2", "bool");
```

Comment:
I think we can have a TODO there, this should be inferrable as `Literal[False]` if we correctly inferred `1 == "a"` as `Literal[False]`

Reply:
also in here!: https://github.com/astral-sh/ruff/pull/13712/files#diff-eff9f162b9d83ddfe43d3f6b28ed022767bc58d12d7ccb0c95d985e827471362R92-R96